libcore/src/SimulationHelper.cpp

```diff
@@ -0,0 +1,218 @@
+    std::map<int, std::shared_ptr<SubRoom>> subroomsConnected;
+    for(const auto & [roomID, room] : building.GetAllRooms()) {
+        std::copy_if(
+            std::begin(room->GetAllSubRooms()),
+            std::end(room->GetAllSubRooms()),
+            std::inserter(subroomsConnected, subroomsConnected.end()),
+            [oldSubRoom](auto const & iterator) -> bool {
+                return iterator.second->IsDirectlyConnectedWith(oldSubRoom);
+            });
+    }
+
+    // check if pedestrian is in one of the neighboring rooms/subrooms
+    auto currentSubRoom = std::find_if(
+        std::begin(subroomsConnected),
+        std::end(subroomsConnected),
+        [&ped](auto const & iterator) -> bool {
+            return iterator.second->IsInSubRoom(ped.GetPos());
+        });
+
+    // pedestrian is in one of the neighboring rooms/subrooms
+    if(currentSubRoom != subroomsConnected.end()) {
+        ped.UpdateRoom(currentSubRoom->second->GetRoomID(), currentSubRoom->second->GetSubRoomID());
+        ped.SetSubRoomUID(currentSubRoom->second->GetUID());
+        return true;
+    } else {
+        ped.UpdateRoom(-1, -1);
+        ped.SetSubRoomUID(-1);
+        return std::nullopt;
+    }
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindPedsReachedFinalGoal(
+    const Building & building,
+    const std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsAtFinalGoal;
+    const auto & goals = building.GetAllGoals();
+
+    std::copy_if(
+        std::begin(peds),
+        std::end(peds),
+        std::inserter(pedsAtFinalGoal, std::end(pedsAtFinalGoal)),
+        [goals](auto const ped) -> bool {
+            return ped->GetFinalDestination() != FINAL_DEST_OUT &&
+                   goals.at(ped->GetFinalDestination())->Contains(ped->GetPos()) &&
+                   goals.at(ped->GetFinalDestination())->GetIsFinalGoal();
+        });
+    return pedsAtFinalGoal;
+}
+
+std::tuple<std::vector<Pedestrian *>, std::vector<Pedestrian *>>
+SimulationHelper::UpdateLocations(const Building & building, const std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsNotRelocated;
+    std::vector<Pedestrian *> pedsChangedRoom;
+
+    // Check for peds, where relocation failed
+    for(auto const ped : peds) {
+        if(auto relocated = UpdateRoom(building, *ped); relocated.has_value()) {
+            if(relocated.value() && ped->ChangedRoom()) {
+                pedsChangedRoom.push_back(ped);
+            }
+        } else {
+            pedsNotRelocated.push_back(ped);
+        }
+    }
+
+    return {pedsChangedRoom, pedsNotRelocated};
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindOutsidePedestrians(
+    const Building & building,
+    std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsOutside;
+    std::vector<Pedestrian *> newPeds;
+
+    std::partition_copy(
+        std::make_move_iterator(std::begin(peds)),
+        std::make_move_iterator(std::end(peds)),
```

Comment:
Why do you use move here?

Reply:
Misunderstanding from my side. Changed it.